rust/src/ikev1/parser.rs

```diff
@@ -0,0 +1,732 @@
+    GroupRandomEcp192,
+    GroupRandomEcp224,
+    GroupBrainpoolEcp224,
+    GroupBrainpoolEcp256,
+    GroupBrainpoolEcp384,
+    GroupBrainpoolEcp512,
+    // Life Type
+    LifeTypeSeconds,
+    LifeTypeKilobytes
+}
+
+impl fmt::Display for AttributeValue {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+            write!(f, "{:?}", self)
+        }
+}
+
+pub struct SaAttribute {
+    pub attribute_format: u8,
+    pub attribute_type: AttributeType,
+    pub attribute_value: AttributeValue,
+    pub numeric_value: Option<u32>,
+    pub hex_value: Option<String>
+}
+
+named! {pub parse_isakmp_header<IsakmpHeader>,
+    do_parse!(
+        init_spi: be_u64 >>
+        resp_spi: be_u64 >>
+        np:       be_u8 >>
+        vers: bits!(
+            tuple!(take_bits!(4u8),take_bits!(4u8))
+        ) >>
+        ex:       be_u8 >>
+        flags:    be_u8 >>
+        id:       be_u32 >>
+        l:        be_u32 >>
+        (
+            IsakmpHeader {
+                init_spi,
+                resp_spi,
+                next_payload: np,
+                maj_ver: vers.0,
+                min_ver: vers.1,
+                exch_type: ex,
+                flags,
+                msg_id: id,
+                length: l,
+            }
+        )
+    )
+}
+
+pub fn parse_security_association(
+    i: &[u8],
+    length: u16,
+) -> IResult<&[u8], SecurityAssociationPayload> {
+    do_parse!(
+        i,
+        domain_of_interpretation: be_u32 >>
+        situation: cond!(domain_of_interpretation == 1, take!(4)) >>
+        data: cond!(domain_of_interpretation == 1, take!(length - 8)) >>
+        (
+            SecurityAssociationPayload {
+                domain_of_interpretation: domain_of_interpretation,
+                situation: situation,
+                data: data
+            }
+        )
+    )
+}
+
+pub fn parse_key_exchange(
+    i: &[u8],
+    length: u16,
+) -> IResult<&[u8], KeyExchangePayload> {
+    do_parse!(
+        i,
+        key_exchange_data: take!(length) >>
+        (
+            KeyExchangePayload {
+                key_exchange_data: key_exchange_data
+            }
+        )
+    )
+}
+
+pub fn parse_proposal(
+    i: &[u8],
+    length: u16,
+) -> IResult<&[u8], ProposalPayload> {
+    do_parse!(
+        i,
+        proposal_number: be_u8 >>
+        proposal_type: be_u8 >>
+        spi_size: be_u8 >>
+        number_transforms: be_u8 >>
+        spi: take!(spi_size) >>
+        data: take!((length - 4) - spi_size as u16) >>
+        (
+            ProposalPayload {
+                proposal_number,
+                proposal_type,
+                spi_size,
+                number_transforms,
+                spi,
+                data
+            }
+        )
+    )
+}
+
+pub fn parse_transform(
+    i: &[u8],
+    length: u16,
+) -> IResult<&[u8], TransformPayload> {
+    do_parse!(
+        i,
+        transform_number: be_u8 >>
+        transform_type: be_u8 >> be_u16 >>
```

Comment:
That `be_u16 >>` is badly indented, and easily mistaken (is the field really ignored?)

Reply:
yes, according to the rfc 2408: 
RESERVED2 (2 octets) - Unused, set to 0.